# Exploratory Data Analysis – Airline Flight Delay Dataset

## Week 1 EDA

This notebook contains the initial exploration of the 2024 airline flight delay dataset. The aim is to understand the dataset, identify important variables, and find initial patterns that can be useful for the flight delay dashboard.

## 1. Dataset Overview

The dataset contains flight information for the year 2024, including airline details, departure and arrival information, flight delays, cancellations, and different causes of delays.

The sample used for initial exploration contains **10,000 rows and 35 columns**.

In [19]:
import pandas as pd
df = pd.read_csv("../data/raw/flight_data_2024_sample.csv")


In [20]:
df.head()

,year,month,day_of_month,day_of_week,fl_date,op_unique_carrier,op_carrier_fl_num,origin,origin_city_name,origin_state_nm,...,diverted,crs_elapsed_time,actual_elapsed_time,air_time,distance,carrier_delay,weather_delay,nas_delay,security_delay,late_aircraft_delay
0,2024,4,18,4,2024-04-18,MQ,3535.0,DFW,"Dallas/Fort Worth, TX",Texas,...,0,151.0,144.0,119.0,835.0,0,0,0,0,0
1,2024,1,1,1,2024-01-01,AA,148.0,CLT,"Charlotte, NC",North Carolina,...,0,286.0,273.0,253.0,1773.0,0,0,0,0,0
2,2024,12,12,4,2024-12-12,9E,5440.0,CHA,"Chattanooga, TN",Tennessee,...,0,59.0,50.0,29.0,106.0,0,0,0,0,0
3,2024,4,8,1,2024-04-08,WN,1971.0,OMA,"Omaha, NE",Nebraska,...,0,180.0,177.0,163.0,1099.0,0,0,0,0,0
4,2024,2,16,5,2024-02-16,WN,862.0,BWI,"Baltimore, MD",Maryland,...,0,90.0,96.0,76.0,399.0,0,0,0,0,0


In [21]:
df.columns

Index(['year', 'month', 'day_of_month', 'day_of_week', 'fl_date',
       'op_unique_carrier', 'op_carrier_fl_num', 'origin', 'origin_city_name',
       'origin_state_nm', 'dest', 'dest_city_name', 'dest_state_nm',
       'crs_dep_time', 'dep_time', 'dep_delay', 'taxi_out', 'wheels_off',
       'wheels_on', 'taxi_in', 'crs_arr_time', 'arr_time', 'arr_delay',
       'cancelled', 'cancellation_code', 'diverted', 'crs_elapsed_time',
       'actual_elapsed_time', 'air_time', 'distance', 'carrier_delay',
       'weather_delay', 'nas_delay', 'security_delay', 'late_aircraft_delay'],
      dtype='str')

In [22]:
df.shape

(10000, 35)

## 2. Airline Distribution

The distribution of flights across different airlines was examined to understand how the sample is represented across carriers.

In [23]:
df['op_unique_carrier'].value_counts()

op_unique_carrier
WN    2022
DL    1459
AA    1385
OO    1080
UA    1033
YX     434
MQ     406
NK     361
B6     331
OH     326
AS     323
F9     295
9E     278
G4     172
HA      95
Name: count, dtype: int64

### Observation

WN has the highest number of flights in the sample, followed by DL, AA, OO, and UA. The number of flights varies across the different carriers.

## 3. Cancellation Analysis

The cancellation status of flights was examined to understand the proportion of cancelled and non-cancelled flights in the sample.

In [24]:
df['cancelled'].value_counts()

cancelled
0    9878
1     122
Name: count, dtype: int64

### Observation

Out of 10,000 flights in the sample, 9,878 were not cancelled and 122 were cancelled. This shows that cancellations form a small portion of the sample.

In [25]:
(df['dep_delay'] > 0).value_counts()

dep_delay
False    6337
True     3663
Name: count, dtype: int64

## 4. Departure Delay Analysis

The sample was examined to identify the number of flights with and without recorded departure delays.

In [26]:
df.loc[df['dep_delay'].notna(), 'dep_delay'].gt(0).value_counts()

dep_delay
False    6221
True     3663
Name: count, dtype: int64

### Observation

Among the 9,884 flights with recorded departure-delay values, 3,663 flights had a positive departure delay, while 6,221 flights did not have a positive departure delay. This indicates that a considerable portion of the recorded flights experienced departure delays.

In [27]:
df[df['dep_delay'] > 0].groupby('op_unique_carrier').size().sort_values(ascending=False)

op_unique_carrier
WN    956
AA    586
DL    487
UA    364
OO    291
NK    145
B6    141
MQ    118
OH    116
AS    115
F9    111
YX     76
G4     64
9E     56
HA     37
dtype: int64

## 5. Airline-wise Delay Rate

The departure delay rate was compared across airlines to identify differences in delay frequency between carriers.

In [28]:
df.groupby('op_unique_carrier')['dep_delay'].apply(lambda x: (x > 0).mean() * 100).sort_values(ascending=False)

op_unique_carrier
WN    47.279921
B6    42.598187
AA    42.310469
NK    40.166205
HA    38.947368
F9    37.627119
G4    37.209302
AS    35.603715
OH    35.582822
UA    35.237173
DL    33.379027
MQ    29.064039
OO    26.944444
9E    20.143885
YX    17.511521
Name: dep_delay, dtype: float64

### Observation

In the sample, WN has the highest departure-delay rate at about 47.3%, while YX has the lowest at about 17.5%. The results show noticeable differences in delay rates across airlines.

In [29]:
df['dep_delay'].describe()

count    9884.000000
mean       13.002428
std        53.614819
min       -22.000000
25%        -6.000000
50%        -2.000000
75%         9.250000
max      2011.000000
Name: dep_delay, dtype: float64

## 6. Monthly Flight Distribution

The number of flights in the sample was examined across the 12 months of 2024 to understand how the sample is distributed over time.

In [30]:
df['month'].value_counts().sort_index()

month
1     744
2     716
3     836
4     856
5     833
6     845
7     917
8     867
9     836
10    946
11    798
12    806
Name: count, dtype: int64

### Observation

The sample contains flights from all 12 months of 2024. The number of flights varies across months, with October having the highest representation and February having the lowest.

In [31]:
df['origin'].value_counts().head(10)

origin
ATL    496
DFW    438
DEN    412
ORD    391
CLT    319
PHX    289
LAX    282
LAS    236
LGA    235
MCO    224
Name: count, dtype: int64